NETOBSERV-2071 CLI use cases for 1.8 (#6)
netobserv/netobserv.github.io: content/posts/2025-01-28-cli_use_cases.md

        | .' / -_)  _/ _ \ '_ (_-</ -_) '_\ V / | (__| |__ | | 
        |_|\_\___|\__\___/_.__/__/\___|_|  \_/   \___|____|___|

------------------------------------------------------------------------
INFO[0000] Log level: info
Option(s): enable_all|peer_ip=10.129.0.48 
INFO[0000] Kernel version: 5.14.0-427.50.1.el9_4.x86_64 
INFO[0000] Starting Flow Capture...                     
INFO[0000] Creating database...                         
INFO[0000] flows.db created                             
INFO[0000] Create flows table...                        
INFO[0000] flows table created 
```

At this stage, the collector wait for incoming data. If nothing shows yet, it means that no traffic is captured. Try to open the route of your application or update the filters of the capture.

Once some traffic is captured, the output will look like:
![cli dropped]({page.image('cli/connectivity-scenario-cli-dropped.png')})

You can cycle to different views using left / right arrow keys and change the displayed enrichment colomns using page up / down ones.
Also, to adapt to your screen height, you can increase / decrease the number of displayed flows using up / down arrow keys.

> In this capture, we see that the traffic is blocked by OVS since it reports the `OVS_DROP_LAST_ACTION` drop cause. This probably means that a Network Policy is involved.

Reviewer: This stuff nolonger available with network events support

Author: ok I can rerun this scenario and see how it goes. I'm supposed to get both drop info + event here ?

Author: That means this will not displayed in [overview drop cause graph](https://github.com/netobserv/network-observability-console-plugin/blob/main/web/src/utils/overview-panels.ts#L71) anymore !

We may need a new graph for events here

Reviewer: we will get still pkt drops but not for ovn netpol

Author: I actually don't get anything in the drops info in that case
![Screenshot from 2025-02-03 11-32-07](https://github.com/user-attachments/assets/0ffb23d5-63dc-47ca-8fe4-47df2713af4e)

Whereas the events are correct:
![Screenshot from 2025-02-03 11-31-45](https://github.com/user-attachments/assets/ca4122a2-2425-4dda-be88-8dd35d416ff7)

It's really confusing in terms of UX

Author: https://issues.redhat.com/browse/NETOBSERV-2092

Author: listed the two cases in https://github.com/netobserv/netobserv.github.io/pull/6/commits/152644b7a8ba8dfe9a9ec6cd0f8b8b92cbf47626

cc @jotak we should be good to publish now :smile_cat:

---
layout: :theme/post
title: "Network Observability On Demand Use Cases"
description: Command line interface usage for concrete scenarios
tags: CLI,Monitoring,Troubleshooting
authors: [jpinsonneau]
---

> # Network Observability On Demand use cases

Author: ```suggestion
# Network Observability On Demand use cases

_Thanks to: Mohamed Mahmoud, Joël Takvorian and Sara Thomas for reviewing_

```

```sh
------------------------------------------------------------------------
         _  _     _       _                       ___ _    ___
        | \| |___| |_ ___| |__ ___ ___ _ ___ __  / __| |  |_ _|
        | .' / -_)  _/ _ \ '_ (_-</ -_) '_\ V / | (__| |__ | | 
        |_|\_\___|\__\___/_.__/__/\___|_|  \_/   \___|____|___|

------------------------------------------------------------------------
INFO[0000] Log level: info
Option(s): enable_all|peer_ip=10.129.0.19 
INFO[0000] Kernel version: 5.14.0-427.50.1.el9_4.x86_64 
INFO[0000] Starting Flow Capture...                     
INFO[0000] Creating database...                         
INFO[0000] flows.db created                             
INFO[0000] Create flows table...                        
INFO[0000] flows table created 
```

At this stage, the collector wait for incoming data. If nothing shows yet, it means that no traffic is captured. Try to open the route of your application or update the filters of the capture.

Once some traffic is captured, the output will look like:
![cli network events]({page.image('cli/connectivity-scenario-cli-events.png')})

> Cycle to the **network events** view. In this capture, we see that the traffic is blocked by a network policy since it reports the `NetpolNamespace` event.

Reviewer 2: [This PR](https://github.com/netobserv/network-observability-cli/pull/135) should provide a better output for the network events

Author: Awesome ! Do you have a screen ?
Else I will spin up a cluster

Reviewer 2: not yet, cluster not started yet

Edit your network policies and give another try.

Behind the scenes in our scenario, we used to have a deny all on the pod label:
```yaml
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: deny-nodejs
  namespace: sample-app
spec:
  podSelector:
    matchLabels:
      app: nodejs-basic
  policyTypes:
    - Ingress
    - Egress
```

Once you updated your policies, you can give another try to your route until you fix the issue:
![cli traffic]({page.image('cli/connectivity-scenario-cli-traffic.png')})

> The network event disappear and your route should open correctly now. On top of that, you can ensure that the Round Trip Time is correct. 

Reviewer 3: `disappears`

> If you are still experienting issues with the route, you may update / get rid of the filter(s) and play with live filtering.

Reviewer 3: experiencing spelling wrong

Author: ```suggestion
If you are still experiencing issues with the route, you may update / get rid of the filter(s) and play with live filtering.
```

![cli live filter]({page.image('cli/connectivity-scenario-cli-live-filter.png')})

- And add as many keywords as you want by pressing enter:
![cli live filters]({page.image('cli/connectivity-scenario-cli-live-filters.png')})

- If your deployment involves some DNS resolution, you can switch to the **DNS** display and show latencies and errors:
Once you updated your policies, you can give another try to your route until you fix the issue:
![cli dns]({page.image('cli/connectivity-scenario-cli-dns.png')})

  Those `NxDomain` errors could be from a misconfiguration in your app deployment trying to reach the wrong domain.

- To check which network is involved, you can switch to **Network Name** enrichment and **UDN** display.
This is showing you which network name, interfaces, direction and User Defined Network name involved.
![cli network]({page.image('cli/connectivity-scenario-cli-network.png')})

Once you are done, simply press `CTRL + C` to exit. Your capture is copied to your local machine for post mortem analysis.

That's the end of the first scenario about connectivity checks !

> ## User Defined Network (Tech Preview)

Reviewer 3: Is this tech preview or dev preview? UDN is dev preview right

> As tech preview, you can enrich flows to get User Defined Network (UDNs) information. To do so, you can use the `--enable_udn_mapping` option.

Reviewer 3: ```suggestion
As developer preview, you can enrich flows to get User Defined Network (UDNs) information. To do so, you can use the `--enable_udn_mapping` option.
```

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: pod-a
  namespace: user-defined-network
...
  annotations:
...
    k8s.v1.cni.cncf.io/network-status: |-
      [{
          "name": "ovn-kubernetes",
          "interface": "eth0",
          "ips": [
              "10.131.0.35"
          ],
          "mac": "0a:58:0a:83:00:23",
          "dns": {}
      },{
          "name": "ovn-kubernetes",
          "interface": "ovn-udn1",
          "ips": [
              "10.0.0.4"
          ],
          "mac": "0a:58:0a:00:00:04",
          "default": true,
          "dns": {}
      }]
...
```

Note that `eth0` is the default pod network and `ovn-udn1`is the User Defined Network. We are going to filter on its IP: `10.0.0.4`.
As this IP is not unique across the cluster, we can add a filter on the port `8080` which will be used in this example and the node annotation `kubernetes.io/hostname=ci-ln-cfqkhfb-72292-6l8l5-worker-c-lrr4q` matching the node running this pod.

All together, it will be like:
```sh
oc netobserv flows --enable_udn_mapping --peer_ip=10.0.0.4 --port=8080 --node-selector=kubernetes.io/hostname:ci-ln-cfqkhfb-72292-6l8l5-worker-c-lrr4q
```

> Once this is running, it will hang for flows to come. In this scenario, we simply run a curl between two pods under the same UDN.

Reviewer 4: Can we clarify this sentence a little like this?

> Once this is running, it will hang for flows to come. In this scenario, we simply run a curl between two pods under the same UDN.

Reviewer 4: I'm not sure about hang, thinking about global English readers. Hang may imply that its stuck, but waiting implies that its depending on something first before it goes about its command.

You can cycle to **UDN mapping** display once you get your flows to see which UDN is involved.

![udns]({page.image('cli/udns.png')})

You can retreive the namespace `user-defined-network` and the name `primary-udn` from the UDN labels column.

## Identify unknown traffic content

In this next scenario, we are going to dig into unknown packets using unencrypted traffic on `TCP` protocol on port `80`. We usually avoid such usage on production cluster and would like to understand who's involved in these.

For the demo purepose, we are using again a nodejs basic deployment, exposing unencrypted traffic on port 80 in `packet-capture-scenario` namespace.

Let's start a packet capture using the following command:
```sh
oc netobserv packets --protocol=TCP --port=80
```

> Similarly to the previous scenario, the script will connect to your cluster and start deploying eBPF agents and collector pod but capturing full packet content this time.

Reviewer 4: ```suggestion
Similarly to the previous scenario, the script connects to your cluster and starts deploying the eBPF agents and collector pod but capturing full packet content this time.
```

Reviewer 4: Globally, you may want to consider revising the future tense "will" and make the verbs present tense like this. It situates the reader more in the now.

Once first packet is captured, you see a table with the amount of packets and bytes:

![cli packets enriched]({page.image('cli/packet-capture-scenario-table-enriched.png')})

You can cycle between the view to see IPs and ports instead of resources:

![cli packets ips]({page.image('cli/packet-capture-scenario-table-ips.png')})

Let the capture run enough time to get some packets and press `CTRL + C` to exit.
Open the generated pcap file using Wireshark and look for your IP as source for example:

![wireshark unencrypted packets]({page.image('cli/wireshark-unencrypted-pcap.png')})

This time, we can see the HTTP content and see the request URL and response. Since we have the entire payload, we cand check for any authentification header, cookie, encoding and so on that may be wrong. 
This is useful to identify traffic content but also to debug a deployment that may be misconfigured or requesting unexpected content on unencrypted connections.

## Encrypted traffic

What about encrypted traffic ? 

> Most of the time, your traffic is be secured and you do not have access to the data it holds. However, Wireshark has the ability to decrypt the content if you [provide the proper secrets](https://wiki.wireshark.org/TLS#tls-decryption). However, getting these secrets may be tricky depending on where the encryption happens and needs some juggling. 

Reviewer 3: instead of `is`, `should`

Author: ```suggestion
Most of the time, your traffic should be secured and you do not have access to the data it holds. However, Wireshark has the ability to decrypt the content if you [provide the proper secrets](https://wiki.wireshark.org/TLS#tls-decryption). However, getting these secrets may be tricky depending on where the encryption happens and needs some juggling. 
```

> Most of the time, your traffic will be secured and you will not have access to the data it holds. However, Wireshark has the ability to decrypt the content if you [provide the proper secrets](https://wiki.wireshark.org/TLS#tls-decryption). However, getting these secrets may be tricky depending on where the encryption happens and needs some juggling. 

Reviewer 4: ```suggestion
Most of the time, your traffic is be secured and you do not have access to the data it holds. However, Wireshark has the ability to decrypt the content if you [provide the proper secrets](https://wiki.wireshark.org/TLS#tls-decryption). However, getting these secrets may be tricky depending on where the encryption happens and needs some juggling. 
```

If you are looking for the traffic between your browser and the cluster, you can simply run the following command to capture [pre-master-secret keylog file](https://wiki.wireshark.org/TLS#using-the-pre-master-secret):
```sh
SSLKEYLOGFILE="keylog.txt" chromium-browser &
```

This generated file contains each client and server traffic secrets to decode your packets.

Then in Wireshark, specify the file path in `Edit→Preferences→Protocols→TLS→(Pre)-Master-Secret log filename`. A single file can hold keys for multiple TLS connections.

![wireshark tls keylog]({page.image('cli/wireshark-tls-keylog.png')})

You should now have access to decrypted packets !

What about a secure connection between two pods ? 

> For these cases, you need to have a good understanding of what's involved behind the scene. If your application rely on OpenSSL, GnuTLS or NSS TLS library, you should be able to set the `SSLKEYLOGFILE` enviroment variable and extract the result.

Reviewer 4: ```suggestion
For these cases, you need to have a good understanding of what's involved behind the scene. If your application relies on OpenSSL, GnuTLS or NSS TLS library, you should be able to set the `SSLKEYLOGFILE` enviroment variable and extract the result.
```

For our nodejs example, you need to either use the cli option [--tls-keylog](https://nodejs.org/api/cli.html#cli_tls_keylog_file) or modify the code and subscribe on [keylog event](https://nodejs.org/api/tls.html#event-keylog). When multiple teams are involved, that may take a while !

> Another alternative is to get a man in the middle approach using a proxy. This can be done using [kubetap](https://soluble-ai.github.io/kubetap/).

Reviewer 4: ```suggestion
Another alternative is a man-in-the-middle approach using a proxy. This can be done using [kubetap](https://soluble-ai.github.io/kubetap/).
```

You can install it using krew by running:
```sh
oc krew install tap
```

Then you can run the following command to port forward the traffic of your service to the proxy:
```sh
oc tap on -n packet-capture-scenario -p 3001 --https nodejs-basic --port-forward
```

The output should look like this:
```sh
Establishing port-forward tunnels to Service...
                                                                                     

Port-Forwards:

  mitmproxy - http://127.0.0.1:2244
  nodejs-basic - http://127.0.0.1:4000
```

Open the mitm proxy url and generate a query to your route. You can see queries listed in the proxy web UI with the content displayed:

![mitm proxy]({page.image('cli/mitm-proxy.png')})

## Feedback
We hope you enjoyed this article!

> Netobserv is an OpenSource project [available on github](https://github.com/netobserv).

Reviewer 2: nit: open source is 2 words

```suggestion
Netobserv is an open source project [available on github](https://github.com/netobserv).
```

---
layout: :theme/post
title: "Network Observability On Demand Use Cases"
description: Command line interface usage for concrete scenarios
tags: CLI,Monitoring,Troubleshooting
authors: [jpinsonneau]
---

# Network Observability On Demand use cases

_Thanks to: Mohamed Mahmoud, Joël Takvorian, Sara Thomas and Amogh Rameshappa Devapura for reviewing_

> If you haven't read it yet, take a look at [CLI 1.8 update](./2025-01-23-cli_whats_new_1.8.md) article to list the new introduced features in release 1.8.

Author: ```suggestion
If you haven't read it yet, take a look at [CLI 1.8 update](https://netobserv.io/posts/network-observability-on-demand-1-8-update/) article to list the new introduced features in release 1.8.
```

Reviewer 2: can you try ../network-observability-on-demand-1-8-update/ ?

Author: ```suggestion
If you haven't read it yet, take a look at [CLI 1.8 update](../network-observability-on-demand-1-8-update/) article to list the new introduced features in release 1.8.
```
